flask_dance/contrib/reddit.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import unicode_literals
+
+from flask_dance.consumer import OAuth2ConsumerBlueprint, OAuth2Session
+from functools import partial
+from flask.globals import LocalProxy, _lookup_app_object
+
+from flask_dance import __version__ as _flask_dance_version
+
+try:
+    from flask import _app_ctx_stack as stack
+except ImportError:
+    from flask import _request_ctx_stack as stack
+
+__maintainer__ = "Sergey Storchay <r8@r8.com.ua>"
+
+
+def _default_user_agent():
+    return "Flask-Dance/%s" % _flask_dance_version
+
+
+class RedditOAuth2Session(OAuth2Session):
+    def __init__(self, *args, **kwargs):
+        super(RedditOAuth2Session, self).__init__(*args, **kwargs)
+
+        # Reddit requires to set non-generic user agent
+        self.headers["User-Agent"] = self.blueprint.user_agent or _default_user_agent()
+
+    def fetch_token(self, *args, **kwargs):
+        # Pass client_id to session so it could trigger Basic Auth
+        return super(RedditOAuth2Session, self).fetch_token(
+            client_id=self.blueprint.client_id, *args, **kwargs)
```

Comment:
Can you explain why this is necessary? I'm not familiar with the Reddit API, so I don't know what their requirements are.

Reply:
Again, Flask (and Python in general) is not my usual stack, so I might be missing something or using non-optimal way to do the things.

Reddit fetch token API is protected by Basic Auth. I checked `requests_oauthlib` and saw that Basic Auth can be triggered by passing user/password or just client_id to `fetch_token` method.

https://github.com/requests/requests-oauthlib/blob/c61c6fbe19908df5b8bf5d8caacde3448f2e0c49/requests_oauthlib/oauth2_session.py#L201

Obvious way would be to pass it with `token_url_params` through `OAuth2ConsumerBlueprint` constructor.
But if we aren't passing client_id/client_secret to `make_reddit_blueprint` and instead are taking it from config, then we don't know client_id during `OAuth2ConsumerBlueprint` initialization.

Hope that makes sense.

So the easiest way for me was to override `fetch_token` and take client_id from Blueprint. Because at the moment of `fetch_token` call we should know for sure client_id (either from initialization or Flask config).

It was a quick hack, so I'm open for suggestions about more elegant solution :)